### PR TITLE
add reward pallet modifications

### PR DESCRIPTION
**Pull request by kwhuo68**
**Summary of changes**
Changes introduced in this pull request:
-Scaffold for pallet modifications for rewards (AP VAnchor updating, etc)

### CODE REVIEW

**Comment by drewstone** [comment]
@saiakilesh good to merge if you review and give it the go ahead.

File: pallets/anonymity-mining/src/lib.rs [hunk 1]
```diff
@@ -192,6 +305,171 @@ pub mod pallet {
...
+	/// Update vanchor
+	fn update_vanchor(
+		tree_id: T::TreeId,
+		merkle_root: T::Element,
+		src_resource_id: ResourceId,
+		latest_leaf_index: T::LeafIndex,
+	) -> DispatchResultWithPostInfo {
+		let src_chain_id: T::ChainId = src_resource_id.typed_chain_id().chain_id().into();
+		if T::VAnchor::has_edge(tree_id, src_chain_id) {
+			T::VAnchor::update_edge(
```
Comment (saiakilesh) [inline pallets/anonymity-mining/src/lib.rs:353]: The APVAnchor is stored only on Substrate. It does not need its edges to be updated.

File: pallets/anonymity-mining/src/lib.rs [hunk 2]
```diff
@@ -192,6 +305,171 @@ pub mod pallet {
...
+		src_tree_id: T::TreeId,
+		dest_tree_id: T::TreeId,
+		src_root_index: T::RootIndex,
+		dest_root_index: T::RootIndex,
+		src_deposit_root: T::Element,
+		src_withdraw_root: T::Element,
+		dest_deposit_root: T::Element,
+		dest_withdraw_root: T::Element,
+	) {
+		// Update roots
```
Comment (saiakilesh) [inline pallets/anonymity-mining/src/lib.rs:384]: I don't understand why we are updating the deposit / withdraw tree roots twice for a `src` and `dest`. Can you please explain? In my opinion, we should have two functions, one for inserting into the cached deposit root storage and one for inserting into the cached withdraw root storage.

File: pallets/anonymity-mining/src/lib.rs [hunk 3]
```diff
@@ -192,6 +305,171 @@ pub mod pallet {
...
+
+		NextDepositRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+		NextWithdrawRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+
+		// Update roots - TODO: modify
+		Self::update_roots(
```
Comment (saiakilesh) [inline pallets/anonymity-mining/src/lib.rs:440]: Why are you updating the cached deposit/withdraw tree roots when claiming AP?

File: pallets/anonymity-mining/src/lib.rs [hunk 4]
```diff
@@ -192,6 +305,171 @@ pub mod pallet {
...
+			dest_deposit_root,
+			dest_withdraw_root,
+		);
+
+		// Update rewards VAnchor
+		let vanchor =
+			Self::update_vanchor(source_tree_id, merkle_root, src_resource_id, latest_leaf_index)?;
+
+		// Handle proof verification
+		Self::handle_proof_verification(&proof_data)?;
```
Comment (saiakilesh) [inline pallets/anonymity-mining/src/lib.rs:456]: Shouldn't this line be at the beginning of the function? We have to check the proof first, before doing anything else.

File: pallets/anonymity-mining/src/lib.rs [hunk 5]
```diff
@@ -192,6 +305,171 @@ pub mod pallet {
...
+		merkle_root: T::Element,
+		latest_leaf_index: T::LeafIndex,
+		proof_data: ProofData<T::Element>,
+		src_deposit_root: T::Element,
+		src_withdraw_root: T::Element,
+		dest_deposit_root: T::Element,
+		dest_withdraw_root: T::Element,
+	) -> DispatchResultWithPostInfo {
+		// Transfer AP tokens to recipient
+		<T as Config<I>>::Currency::transfer(
```
Comment (saiakilesh) [inline pallets/anonymity-mining/src/lib.rs:406]: Claiming AP only consists of inserting a commitment into the APVAnchor, we shouldn't transfer any tokens.

**Review by saiakilesh — COMMENTED** [review]
Along, with my comments on specific lines of code, here are a few overarching things:
1. We have renamed things: `DepositTree` -> `UnspentTree`, `WithdrawTree -> SpentTree`.
1. I would rather separate concerns and have two different pallets, a claims pallet and an AMM pallet. The claims pallet is responsible for storing reward tree roots, checking claims circuit proof, and inserting commitments into the AP VAnchor. The AMM pallet is responsible for converting AP into Tangle tokens (or whatever the reward token is).

File: pallets/anonymity-mining-claims/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+		NextDepositRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+		NextWithdrawRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+
+		// Updating roots, TODO: modify
+		Self::update_deposit_root(deposit_tree_id, deposit_root_index, deposit_root);
```
Comment (saiakilesh) [inline pallets/anonymity-mining-claims/src/lib.rs:365]: Again, not sure why updating deposit/withdraw roots while claiming.

File: pallets/anonymity-mining-claims/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+		Self::update_withdraw_root(withdraw_tree_id, withdraw_root_index, withdraw_root);
+
+		// Flag deposit nullifier as being used, TODO: modify
+		for nullifier in &proof_data.input_nullifiers {
```
Comment (saiakilesh) [inline pallets/anonymity-mining-claims/src/lib.rs:369]: Need to also store the withdraw tree nullifier hashes.

File: pallets/anonymity-mining-claims/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+			Self::add_deposit_nullifier_hash(deposit_tree_id, *nullifier)?;
+		}
+
+		// Insert output commitments into the AP tree
+		for comm in &proof_data.output_commitments {
+			T::LinkableTree::insert_in_order(deposit_tree_id, *comm)?;
```
Comment (saiakilesh) [inline pallets/anonymity-mining-claims/src/lib.rs:375]: Why are you inserting into deposit tree instead of AP VAnchor?

**Review by saiakilesh — CHANGES_REQUESTED** [review]
`deposit -> unspent`, `withdraw -> spent`

**Pull request merged**